org.idempiere.test/src/org/idempiere/test/base/POTest.java

```diff
@@ -449,4 +451,19 @@ public void testOptimisticLocking()
 		msg2 = new MMessage(Env.getCtx(), msg1.getAD_Message_ID(), getTrxName());
 		assertEquals(msg1.getMsgText(), msg2.getMsgText());
 	}
+
+	@Test
+	public void testVirtualColumnLoad() {
+		MTest testPo = new MTest(Env.getCtx(), getClass().getName(), 1);
+		testPo.save();
+
+		// asynchronous (default) virtual column loading
+		BigDecimal expected = new BigDecimal(123.45d).setScale(2, RoundingMode.HALF_DOWN);
+		assertTrue(expected.compareTo(testPo.getTestVirtualQty()) == 0);
+
+		// synchronous virtual column loading
+		testPo = new MTest(Env.getCtx(), testPo.get_ID(), getTrxName(), MTest.COLUMNNAME_TestVirtualQty);
+		assertTrue(expected.compareTo(testPo.getTestVirtualQty()) == 0);

```

Comment:
would be nice if we can add test to assert for both lazy and eager loading of virtual columns (probably need to add method to support such assertion).

Reply:
Right now the test just makes sure that things are working. How the data is being loaded is confirmed by debugger / code coverage.
There could be a hack-ish way to better confirm this by using `PO.get_ValueOld(int)`. What do you think?

Reply:
I guess the straightforward approach is to add method that allow us to query the state of the virtual columns. for e.g, isVirtualColumnLoaded or getVirtualColumnsNotLoaded (return array of virtual columns that's not loaded yet).

Reply:
While I agree on the goal, I believe there's currently no way for testing this properly, without using reflection or some other hack. We could expose and check the set of virtual columns loaded but that wouldn't guarantee that the values were properly loaded. We'd have to peek into `PO.m_oldValues`, which is private.
At time of this writing all tests are passing and code coverage is OK for all relevant paths.
Now if there are no other observations that can prevent the merging of this branch, we're switching our focus to other (unreported) core issues.

Reply:
m_oldValues is private but you can use get_ValueOld method to access it. The current test doesn't cover the difference between PO(ctx, id, trxName) and PO(ctx, id, trxName, VirtualColumns ...) so it would be nice if we can get that cover as well.